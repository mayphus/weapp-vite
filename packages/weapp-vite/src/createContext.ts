import type { LoadConfigOptions } from './context'
import { getCompilerContext } from './context/getInstance'

export async function createCompilerContext(options?: Partial<LoadConfigOptions & { key?: string }>) {
  // 先初始化 ConfigService
  const ctx = getCompilerContext(options?.key)
  const { configService, scanService } = ctx
  await configService.load(options)
  // prefilght
  try {
    await scanService.loadAppEntry()
  }
  catch {
    // prefilght catch
  }

  return ctx
}
