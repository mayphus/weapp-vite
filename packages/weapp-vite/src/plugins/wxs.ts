import type { CompilerContext } from '@/context'
import type { WxmlDep } from '@/types'
import type { EmittedFile, PluginContext } from 'rollup'
import type { Plugin } from 'vite'
import { jsExtensions } from '@/constants'
import { transformWxsCode } from '@/wxs'
import { removeExtension } from '@weapp-core/shared'
import fs from 'fs-extra'
import { LRUCache } from 'lru-cache'
import path from 'pathe'

export const wxsCodeCache = new LRUCache<string, string>(
  {
    max: 512,
  },
)

export function wxs({ configService, wxmlService }: CompilerContext): Plugin[] {
  const wxsMap = new Map<string, {
    emittedFile: EmittedFile
  }>()

  async function transformWxs(this: PluginContext, {
    wxsPath,
  }: {
    wxsPath: string
  }) {
    if (await fs.exists(wxsPath)) {
      this.addWatchFile(wxsPath)
      const arr = wxsPath.match(/\.wxs(\.[jt]s)?$/)
      let isRaw = true
      if (arr) {
        isRaw = !arr[1]
      }
      const rawCode = await fs.readFile(wxsPath, 'utf8')
      let code = wxsCodeCache.get(rawCode)

      const dirname = path.dirname(wxsPath)
      if (code === undefined) {
        const { result, importees } = transformWxsCode(rawCode, {
          filename: wxsPath,
        })
        if (typeof result?.code === 'string') {
          code = result.code
        }
        await Promise.all(
          importees.map(({ source }) => {
            return transformWxs.call(this, {
              wxsPath: path.resolve(dirname, source),
            })
          }),
        )
      }

      if (code !== undefined) {
        wxsMap.set(wxsPath, {
          emittedFile: {
            type: 'asset',
            fileName: configService.relativeAbsoluteSrcRoot(isRaw ? wxsPath : removeExtension(wxsPath)),
            source: code,
          },
        })
        wxsCodeCache.set(rawCode, code)
      }
    }
  }
  function handleWxsDeps(this: PluginContext, deps: WxmlDep[], absPath: string) {
    return Promise.all(
      deps.filter(x => x.tagName === 'wxs').map(async (wxsDep) => {
        const arr = wxsDep.value.match(/\.wxs(\.[jt]s)?$/)
        if (jsExtensions.includes(wxsDep.attrs.lang) || arr) {
          const wxsPath = path.resolve(path.dirname(absPath), wxsDep.value)

          await transformWxs.call(this, {
            wxsPath,
          })
        }
      }),
    )
  }

  return [
    {
      name: 'weapp-vite:wxs',
      enforce: 'pre',
      buildStart() {
        wxsMap.clear()
      },

      async buildEnd() {
        await Promise.all(wxmlService.tokenMap.entries().map(([id, token]) => {
          return handleWxsDeps.call(this, token.deps, id)
        }))

        wxsMap.values().forEach(({ emittedFile }) => {
          this.emitFile(emittedFile)
        })
      },
    },
  ]
}
