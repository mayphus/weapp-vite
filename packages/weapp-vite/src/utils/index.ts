export * from './file'
export * from './json'
export * from './projectConfig'
export * from './scan'
// export * from './wxml'
// export * from './wxs'
