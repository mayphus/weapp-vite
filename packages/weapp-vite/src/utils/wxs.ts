import type { BuildOptions } from 'esbuild'
// import type { TsupOptions } from '../types'
import { defu } from '@weapp-core/shared'
import { build as esbuild } from 'esbuild'
// import { build } from 'vite'
// import { build } from 'vite'

export async function buildWxs(options: BuildOptions) {
  // build({
  //   build:{
  //     target:
  //   }
  // })
  // const { build: tsupBuild } = await import('tsup')
  // build({})
  const mergedOptions: BuildOptions = defu<BuildOptions, BuildOptions[]>(options, {
    // entry: wxsPaths,
    format: 'cjs',
    // outDir: ctx.outDir,
    target: 'es5',
    // silent: true,
    // shims: true,
    loader: {
      '.wxs': 'js',
    },
    entryNames: '[dir]/[name]',
    outExtension: {
      '.js': '.wxs',
    },
    sourcemap: false,
    // outbase: 'src',
    // esbuildOptions: (options) => {
    //   // options.entryNames
    //   // options.outExtension
    //   options.entryNames = '[dir]/[dir]/[name]'
    //   options.outExtension = {
    //     '.js': '.wxs',
    //   }
    //   options.assetNames = '[dir]/[name]'
    //   options.chunkNames = '[dir]/[name]'
    //   // 保持名称
    //   // options.keepNames = true
    // },
    // outExtension: ({ format, options }) => {
    //   // console.log(format)
    //   return {
    //     js: '.wxs', // .wxs',
    //   }
    // },
    // config: false,

  })
  await esbuild(mergedOptions)
}
