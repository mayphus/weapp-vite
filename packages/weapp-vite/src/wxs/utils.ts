import { addExtension, removeExtensionDeep } from '@weapp-core/shared'

export function normalizeWxsFilename(value: string) {
  return addExtension(removeExtensionDeep(value), '.wxs')
}
