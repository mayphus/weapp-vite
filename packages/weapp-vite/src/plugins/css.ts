import type { Plugin } from 'vite'
import type { CompilerContext } from '../context'
import { objectHash } from '@weapp-core/shared'
import { LRUCache } from 'lru-cache'
import { cssPostProcess } from '../postcss'
import { changeFileExtension, isJsOrTs } from '../utils'

export const cssCodeCache = new LRUCache<string, string>(
  {
    max: 512,
  },
)

function getCacheKey(code: string, options?: any) {
  return objectHash({
    code,
    options,
  })
}

export function css({ configService }: CompilerContext): Plugin[] {
  return [
    {
      name: 'weapp-vite:css',
      enforce: 'pre',
      async generateBundle(_opts, bundle) {
        const bundleKeys = Object.keys(bundle)
        // 必须这样做，防止 css 相同的情况下，被合并
        await Promise.all(
          bundleKeys.map(async (bundleKey) => {
            const asset = bundle[bundleKey]
            if (asset.type === 'asset') {
              if (bundleKey.endsWith('.css')) {
                // 多个 js 文件 引入同一个样式的时候，此时 originalFileNames 是数组
                await Promise.all(asset.originalFileNames.map(async (originalFileName) => {
                  if (isJsOrTs(originalFileName)) {
                    const fileName = configService.relativeSrcRoot(
                      changeFileExtension(originalFileName, configService.outputExtensions.wxss),
                    )
                    const rawCss = asset.source.toString()
                    const cachekey = getCacheKey(rawCss, { platform: configService.platform })
                    let css = cssCodeCache.get(cachekey)
                    if (!css) {
                      css = await cssPostProcess(
                        rawCss,
                        { platform: configService.platform },
                      )
                      cssCodeCache.set(cachekey, css)
                    }

                    this.emitFile({
                      type: 'asset',
                      fileName,
                      source: css,
                    })
                  }
                }))

                delete bundle[bundleKey]
              }
            }
          }),
        )
      },
    },
  ]
}
