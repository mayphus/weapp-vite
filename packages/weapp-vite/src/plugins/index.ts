import type { Plugin, ResolvedConfig } from 'vite'
import type { CompilerContext } from '../context'
import { addExtension, removeExtension } from '@weapp-core/shared'
import { fdir as Fdir } from 'fdir'
import fs from 'fs-extra'
import MagicString from 'magic-string'
import path from 'pathe'
import { isCSSRequest } from 'vite'
import { supportedCssExtensions } from '../constants'
import { createDebugger } from '../debugger'
import { defaultExcluded } from '../defaults'
import { changeFileExtension, isJsOrTs, resolveJson } from '../utils'
import { getCssRealPath, parseRequest } from './parse'

const debug = createDebugger('weapp-vite:plugin')

// <wxs module="wxs" src="./test.wxs"></wxs>
// https://developers.weixin.qq.com/miniprogram/dev/framework/view/wxml/event.html

// https://developers.weixin.qq.com/miniprogram/dev/reference/wxml/import.html

// https://github.com/rollup/rollup/blob/c6751ff66d33bf0f4c87508765abb996f1dd5bbe/src/watch/fileWatcher.ts#L2
// https://github.com/rollup/rollup/blob/c6751ff66d33bf0f4c87508765abb996f1dd5bbe/src/watch/watch.ts#L174
export function vitePluginWeapp(ctx: CompilerContext): Plugin[] {
  let configResolved: ResolvedConfig

  function relative(p: string) {
    return path.relative(configResolved.root, p)
  }
  function getInputOption(entries: string[]) {
    return entries
      .reduce<Record<string, string>>((acc, cur) => {
        acc[relative(cur)] = cur
        return acc
      }, {})
  }

  return [
    {
      name: 'weapp-vite:pre',
      enforce: 'pre',
      api: {

      },
      // config->configResolved->|watching|options->buildStart
      config(config, env) {
        debug?.(config, env)
      },
      configResolved(config) {
        debug?.(config)
        configResolved = config
      },
      async options(options) {
        await ctx.scanAppEntry()
        const input = getInputOption([...ctx.entriesSet])
        options.input = input
      },
      async buildStart() {
        const { build } = configResolved

        const ignore: string[] = [
          ...defaultExcluded,
        ]

        ignore.push(
          ...[
            `${build.outDir}/**`,
            'project.config.json',
            'project.private.config.json',
            'package.json',
            'tsconfig.json',
            'tsconfig.node.json',
          ],
        )
        const relFiles = await new Fdir()
          .withRelativePaths()
          .globWithOptions(
            [path.join(ctx.srcRoot ?? '', '**/*.{wxml,wxs,png,jpg,jpeg,gif,svg,webp}')],
            {
              cwd: ctx.cwd,
              ignore,
            },
          )
          .crawl(ctx.cwd)
          .withPromise()

        for (const file of relFiles) {
          const filepath = path.resolve(ctx.cwd, file)

          this.addWatchFile(filepath)
          const isMedia = !/\.(?:wxml|wxs)$/.test(file)
          this.emitFile({
            type: 'asset',
            fileName: ctx.relativeSrcRoot(file),
            source: isMedia ? await fs.readFile(filepath) : await fs.readFile(filepath, 'utf8'),
          })
        }
        for (const entry of ctx.entries) {
          if (entry.jsonPath) {
            this.addWatchFile(entry.jsonPath)
            if (entry.json) {
              const fileName = ctx.relativeSrcRoot(path.relative(ctx.cwd, entry.jsonPath))
              this.emitFile({
                type: 'asset',
                fileName,
                source: resolveJson(entry.json),
              })
            }
          }
        }
      },
      resolveId(source) {
        if (/\.wxss$/.test(source)) {
          return source.replace(/\.wxss$/, '.css?wxss')
        }
      },
      async load(id) {
        if (ctx.entriesSet.has(id)) {
          const base = removeExtension(id)
          const ms = new MagicString(fs.readFileSync(id, 'utf8'))
          for (const ext of supportedCssExtensions) {
            const mayBeCssPath = addExtension(base, ext)

            if (fs.existsSync(mayBeCssPath)) {
              this.addWatchFile(mayBeCssPath)
              ms.prepend(`import '${mayBeCssPath}'\n`)
            }
          }

          return {
            code: ms.toString(),
          }
        }
        else if (isCSSRequest(id)) {
          const parsed = parseRequest(id)
          const realPath = getCssRealPath(parsed)
          if (await fs.exists(realPath)) {
            const css = await fs.readFile(realPath, 'utf8')
            return {
              code: css,
            }
          }
        }
      },
      // for debug
      // watchChange(id, change) {
      //   console.log(id, change)
      // },
      generateBundle(_options, bundle) {
        const bundleKeys = Object.keys(bundle)
        for (const bundleKey of bundleKeys) {
          const asset = bundle[bundleKey]
          if (bundleKey.endsWith('.css') && asset.type === 'asset' && typeof asset.originalFileName === 'string' && isJsOrTs(asset.originalFileName)) {
            const newFileName = ctx.relativeSrcRoot(
              changeFileExtension(asset.originalFileName, 'wxss'),
            )
            this.emitFile({
              type: 'asset',
              fileName: newFileName,
              source: asset.source,
            })
            delete bundle[bundleKey]
          }
        }
      },
      // writeBundle(options, bundle) {
      //   console.log(options, bundle)
      // },
    },
    {
      name: 'weapp-vite',
    },
    {
      name: 'weapp-vite:post',
      enforce: 'post',
    },
  ]
}
