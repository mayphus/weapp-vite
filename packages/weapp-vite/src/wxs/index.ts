import babel from '@babel/core'
import * as t from '@babel/types'
import { defu } from '@weapp-core/shared'
import { normalizeWxsFilename } from './utils'

export interface TransformWxsCodeOptions {
  filename?: string
  // inline?: boolean
}

export {
  normalizeWxsFilename,
}

export function transformWxsCode(code: string, options?: TransformWxsCodeOptions) {
  const { filename } = defu<TransformWxsCodeOptions, TransformWxsCodeOptions[]>(options, {
    filename: '_default.wxs',
  })
  return babel.transformSync(code, {
    presets: [
      ['@babel/preset-env'],
      ['@babel/preset-typescript'],
    ],
    parserOpts: {
      plugins: ['typescript'],
    },
    filename,
    plugins: [
      {
        visitor: {
          Directive: {
            enter(p) {
              p.remove()
            },
          },
          CallExpression: {
            enter(p) {
              if (p.get('callee').isIdentifier({
                name: 'require',
              }) && p.get('arguments').length === 1) {
                const importee = p.get('arguments')[0]
                if (importee.isStringLiteral()) {
                  importee.node.value = normalizeWxsFilename(importee.node.value)
                }
              }
            },
          },
          ExpressionStatement(p) {
            const expression = p.node.expression

            if (
              expression.type === 'CallExpression'
              && expression.callee.type === 'MemberExpression'
              && t.isIdentifier(expression.callee.object)
              && expression.callee.object.name === 'Object'
              && t.isIdentifier(expression.callee.property)
              && expression.callee.property.name === 'defineProperty'
              && expression.arguments.length >= 2
              && t.isIdentifier(expression.arguments[0])
              && expression.arguments[0].name === 'exports'
              && t.isStringLiteral(expression.arguments[1])
              && expression.arguments[1].value === '__esModule'
            ) {
              p.remove()
            }
          },
          NewExpression: {
            enter(p) {
              if (p.get('callee').isIdentifier({
                name: 'RegExp',
              })) {
                p.replaceWith(
                  t.callExpression(t.identifier('getRegExp'), p.get('arguments').map(x => x.node)),
                )
              }
              else if (p.get('callee').isIdentifier({
                name: 'Date',
              })) {
                p.replaceWith(
                  t.callExpression(t.identifier('getDate'), p.get('arguments').map(x => x.node)),
                )
              }
            },
          },
          RegExpLiteral: {
            enter(p) {
              const args = [t.stringLiteral(p.node.pattern)]
              if (p.node.flags) {
                args.push(t.stringLiteral(p.node.flags))
              }
              p.replaceWith(
                t.callExpression(t.identifier('getRegExp'), args),
              )
            },
          },
          MemberExpression: {
            enter(p) {
              if (p.get('object').isIdentifier({
                name: 'exports',
              })) {
                p.replaceWith(
                  t.memberExpression(
                    t.memberExpression(t.identifier('module'), t.identifier('exports')),
                    p.get('property').node,
                  ),
                )
              }
            },
          },
        },
      },
    ],
  })
}
