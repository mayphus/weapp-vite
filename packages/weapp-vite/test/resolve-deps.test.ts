import type { CompilerContext } from '@/context'
import CI from 'ci-info'
import fs from 'fs-extra'
import path from 'pathe'
import { createCompilerContext } from '@/createContext'
import { cssCodeCache } from '@/plugins/css'
import { wxsCodeCache } from '@/plugins/wxs'
import { getFixture, scanFiles } from './utils'

describe.skipIf(CI.isCI)('resolve-deps', () => {
  const cwd = getFixture('resolve-deps')
  const distDir = path.resolve(cwd, 'dist')
  let ctx: CompilerContext
  beforeAll(async () => {
    await fs.remove(distDir)
    ctx = await createCompilerContext({
      cwd,
    })
    await ctx.buildService.build()
    expect(await fs.exists(distDir)).toBe(true)
  })

  it('scanFiles', async () => {
    const files = await scanFiles(distDir)
    // expect(files).toMatchSnapshot()
    const codes = await Promise.all(
      files.filter(x => x.endsWith('.json')).map((x) => {
        return fs.readFile(path.resolve(distDir, x), 'utf-8')
      }),
    )
    expect(codes).toMatchSnapshot()
    expect(wxsCodeCache.size).toBe(0)
    expect(cssCodeCache.size).toBe(11)
    expect(ctx.scanService.subPackageMap).toMatchSnapshot()
    expect(await fs.exists(path.resolve(distDir, 'miniprogram_npm'))).toBe(true)
    // expect(await fs.exists(path.resolve(distDir, 'miniprogram_npm/buffer'))).toBe(true)
    // expect(await fs.exists(path.resolve(distDir, 'miniprogram_npm/gm-crypto'))).toBe(true)
  })
})
