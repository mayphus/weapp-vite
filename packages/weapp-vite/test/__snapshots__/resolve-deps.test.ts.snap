// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resolve-deps > scanFiles 1`] = `
[
  "app.js",
  "app.json",
  "app.wxss",
  "pages/button/base/index.js",
  "pages/button/base/index.json",
  "pages/button/base/index.wxml",
  "pages/button/base/index.wxss",
  "pages/button/block-btn/index.js",
  "pages/button/block-btn/index.json",
  "pages/button/block-btn/index.wxml",
  "pages/button/block-btn/index.wxss",
  "pages/button/button.js",
  "pages/button/button.json",
  "pages/button/button.wxml",
  "pages/button/button.wxss",
  "pages/button/disabled/index.js",
  "pages/button/disabled/index.json",
  "pages/button/disabled/index.wxml",
  "pages/button/disabled/index.wxss",
  "pages/button/ghost-btn/index.js",
  "pages/button/ghost-btn/index.json",
  "pages/button/ghost-btn/index.wxml",
  "pages/button/ghost-btn/index.wxss",
  "pages/button/group-btn/index.js",
  "pages/button/group-btn/index.json",
  "pages/button/group-btn/index.wxml",
  "pages/button/group-btn/index.wxss",
  "pages/button/icon-btn/index.js",
  "pages/button/icon-btn/index.json",
  "pages/button/icon-btn/index.wxml",
  "pages/button/icon-btn/index.wxss",
  "pages/button/shape/index.js",
  "pages/button/shape/index.json",
  "pages/button/shape/index.wxml",
  "pages/button/shape/index.wxss",
  "pages/button/size/index.js",
  "pages/button/size/index.json",
  "pages/button/size/index.wxml",
  "pages/button/size/index.wxss",
  "pages/button/skyline/button.js",
  "pages/button/skyline/button.json",
  "pages/button/skyline/button.wxml",
  "pages/button/skyline/button.wxss",
  "pages/button/theme/index.js",
  "pages/button/theme/index.json",
  "pages/button/theme/index.wxml",
  "pages/button/theme/index.wxss",
  "sitemap.json",
  "theme.json",
]
`;

exports[`resolve-deps > scanFiles 2`] = `
[
  "{
  "pages": [
    "pages/button/button",
    "pages/button/skyline/button"
  ],
  "window": {},
  "style": "v2",
  "componentFramework": "glass-easel",
  "sitemapLocation": "sitemap.json"
}",
  "{
  "component": true,
  "styleIsolation": "shared",
  "usingComponents": {}
}",
  "{
  "component": true,
  "usingComponents": {}
}",
  "{
  "navigationBarTitleText": "Button",
  "usingComponents": {
    "base": "./base",
    "size": "./size",
    "icon-btn": "./icon-btn",
    "group-btn": "./group-btn",
    "block-btn": "./block-btn",
    "ghost-btn": "./ghost-btn",
    "shape": "./shape",
    "theme": "./theme",
    "disabled": "./disabled"
  }
}",
  "{
  "component": true,
  "usingComponents": {}
}",
  "{
  "component": true,
  "usingComponents": {}
}",
  "{
  "component": true,
  "usingComponents": {}
}",
  "{
  "component": true,
  "usingComponents": {}
}",
  "{
  "component": true,
  "usingComponents": {}
}",
  "{
  "component": true,
  "usingComponents": {}
}",
  "{
  "navigationStyle": "custom",
  "usingComponents": {
    "base": "../base",
    "size": "../size",
    "icon-btn": "../icon-btn",
    "group-btn": "../group-btn",
    "block-btn": "../block-btn",
    "ghost-btn": "../ghost-btn",
    "shape": "../shape",
    "theme": "../theme",
    "disabled": "../disabled"
  },
  "renderer": "skyline",
  "componentFramework": "glass-easel",
  "disableScroll": true
}",
  "{
  "component": true,
  "usingComponents": {}
}",
  "{
  "desc": "关于本文件的更多信息，请参考文档 https://developers.weixin.qq.com/miniprogram/dev/framework/sitemap.html",
  "rules": [
    {
      "action": "allow",
      "page": "*"
    }
  ]
}",
  "{
  "light": {},
  "dark": {}
}",
]
`;
